Bare bones implementation of cupy data dense without going through cython

MrRobot2211 opened the pull request:
This is a no cython  branch. I managed to inherit from the `data.Data` type, but this leads to some workarounds that not appear quite right.

For example when dealing with the `shape` attribute which is originally defined as read-only at cython level.   

I am not quite sure about the directory organization I chose, but I think it is something to focus on later on.
Tests clean-up, number of repetitions and parameters setting is something that will be solved afterwards, as well.

MrRobot2211 committed:
full loop  test working

MrRobot2211 committed:
stubbed linalg tests

MrRobot2211 committed:
renamed dense class

MrRobot2211 committed:
Merge branch 'main' of https://github.com/qutip/qutip-cupy into dense_no_cython

MrRobot2211 committed:
corrected install requires format

hodgestar commented on .gitignore:
```diff
@@ -6,6 +6,7 @@
 .f2py_f2cmap
 .idea/
 .vscode/
+src/qutip_cupy/*.cpp
```
I usually add project specific ignores like this one right at the end of the `.gitignore` file (so that they're easy to find and clean up later if needed).

jakelishman replied:
you can also add it into the `.gitignore` that's in `src/qutip_cupy`.  If you don't like the multiple `.gitignore`, then you should delete the one in` src/qutip_cupy` to avoid fracturing the concerns.

MrRobot2211 replied:
Good points. I added it in the `.gitignore`  inside  `src/qutip_cupy` at the end of the file. Looked like the right place.

hodgestar commented on src/qutip_cupy/cupy_dense.py:
```diff
@@ -0,0 +1,36 @@
+class CuPyDense(data.Data):
+    def __init__(self, data, shape=None, copy=True):
+        base = cp.array(data, dtype=cp.complex128, order='K', copy=copy)
+        if shape is None:
```
Hopefully a lot of this shape handling code can be removed once https://github.com/qutip/qutip/pull/1584 is merged.

jakelishman replied:
No, all the shape handling will still be required - it's up to base classes to ensure they're passing sensible shapes, because only they know what they should default to given their data.

hodgestar replied:
Good point, but hopefully the property can go away and we can just call `super().__init__(shape)` instead.

jakelishman replied:
Yeah, the extra property won't be necessary any more for sure.

MrRobot2211 replied:
It is removed as an attribute, but kept as an argument. I somewhat changed the if-ifelse shape checks so that reshape is  only enforced when both possible and needed.

hodgestar reviewed (commented):
This is not a complete review, but I think this simple approach is the right way for us to implement qutip-cupy. CuPy already provides functions to efficiently execute CUDA operations from Python so there shouldn't be a lot of need for us to write custom Cython code (although perhaps we will write custom CUDA kernels or fancier Cython code later on).

jakelishman commented on src/qutip_cupy/__init__.py:
```diff
@@ -1 +1,21 @@
 from .version import version as __version__
+
+from qutip.core import data
```
We shouldn't be re-exporting parts of `qutip` from this package's main namespace - be sure to `del` it after you've used it.

MrRobot2211 replied:
Good point.

jakelishman commented on src/qutip_cupy/__init__.py:
```diff
@@ -1 +1,21 @@
+
+def dense_from_new(newarray):
+    
+    dense_np = data.Dense(newarray._cp.tolist())
+    return dense_np
```
This is quite inefficient - you probably want something like `data.Dense(cupy.asnumpy(array), copy=False)`.  Also, it would be better if these functions have descriptive names - what does `new` refer to?

MrRobot2211 replied:
Yeahp I changed the naming of the variables everywhere to make them more descriptive. And changed the casting method to `cp.asnumpy` iit also powert the `toarray`method which was undefined till that point.

jakelishman commented on src/qutip_cupy/cupy_dense.py:
```diff
@@ -0,0 +1,36 @@
+import cupy as cp
+import numbers
+from qutip.core import data
+class CuPyDense(data.Data):
```
This is sort of neither here nor there, but if we're going to commit to pep8, this name should be `CupyDense` (no capital p).

We could also argue that it could just be `Dense`, since it should generally be accessed through qualified access to `qutip_cupy.Dense`, so the extra bit is probably unnecessary.

jakelishman replied:
Actually I may be wrong on this - I was thinking about the "rule" that acronyms should be all capitalised.  Still `CuPyDense` looks very odd to me, because "cupy" is one word, no matter how they stylise it.  It's also rather more annoying to type!

hodgestar replied:
I will put in a small vote for `CupyDense` too, but I'm only +0 on it. I will easily live with `CuPyDense` too. :)

jakelishman commented on tests/test_base.py:
```diff
@@ -4,3 +4,26 @@
+
+def test_conversion_cycle():
+    from qutip_cupy import CuPyDense
+    from qutip_cupy import data
```
This module `data` doesn't belong to `qutip_cupy`!  This is `qutip.data`, it shouldn't even be publically accessible from `qutip_cupy`, let alone used from it.

MrRobot2211 replied:
This was really messy on my part. but I changed it here and with the call to `del` in the `__init__` file  it is solved.

jakelishman commented on tests/test_base.py:
```diff
@@ -4,3 +4,26 @@
+
+    old_dense = data.Dense([[0,1,2,6,7,8]])
+
+    tr1 = data.to[CuPyDense, data.Dense](old_dense)
+
+    assert (old_dense.to_array() == data.to[data.Dense, CuPyDense](tr1).to_array()).all()
```
The normal form of `data.to` is `data.to(array, type)` as a function - it's easier to read, and the function will infer what it needs.

MrRobot2211 replied:
Nice, it looks a lot better this way.

MrRobot2211 committed:
corrects initialization of shape

MrRobot2211 committed:
corrects unnecessary import of qutip data module

MrRobot2211 committed:
minor gitignore refactors

MrRobot2211 committed:
Refactor: Changes the location of
conversion functions
and deletes qutip's data layer after registering

MrRobot2211 committed:
cupy_dense __init__  Avoids unnessary reshapes

MrRobot2211 committed:
Complete basic implementation
of the CuPyDense class

MrRobot2211 committed:
changed calls to Data.to directly as  a function

MrRobot2211 committed:
added a minimal set of tests

MrRobot2211 committed:
wraps  adjoint,transpse and conj methods
and registers them to the dispatchers

MrRobot2211 committed:
tests: changed name and refactor mark as fixture

MrRobot2211 commented:
I am thinking of leaving this PR as it is and adding the rest of the functions ( which are not part of the base `Data` definition) in a separate one.
I am probably missing one minor clean-up.
What are your thoughts about docstrings?  Do we take all the basic functions which are in `Data` as being self explanatory?
In `dev.major` explanation are included for as_array and to_array, but in our case we have only to_array as we copy the memory from `device` to `host` .

MrRobot2211 committed:
adds docstring to to_array

MrRobot2211 commented:
@hodegestar  I have been doing some benchmarking and completing the rest of the functions in the other PR. The aggressive calls to `cp.array`  produces a time penalty of around 1.7X when doing some large matrix multiplications. I hhave two options to circumvent this or I can define methods like this:

```
def adjoint(self):
        
        cp_out = self._cp.transpose().conj()
       
        return build_cpd_from_cp(cp_out)

def  build_cpd_from_cp( data):
      out = CuPyDense.__new__(CuPyDense)
      super(CuPyDense , out).__init__(data.shape)
      out._cp = data
      return out

```

 ( which avoids the shape checks, but this new `build_cpd_from_cp` constructor seems awkward) 
 or by adding a type check in the `__init__`

``` 
def __init__(self, data, shape=None, copy=True):
        if isinstance(data, cp.ndarray):
            shape = data.shape
            self._cp = data
```


What are your thoughts here?

MrRobot2211 committed:
reducing calls to cp.array

leofang commented:
Although I am on vacation now, I can help if someone points me to the relevant part of the code. I only have a few general thoughts (pardon my brevity as I am on my cell):
- favor `cp.asarray` over `cp.array` unless you have a special reason (this is also true for NumPy)
- avoid redundant data transfer: if you already have a GPU array, try making use of it to the maximal extent 
- benchmark against a large size problem: for small problem sizes (say gemm of two 10x10 matrices), CPU almost always wins, so don't waste your time.

MrRobot2211 committed:
CuPydense_init  : adds copy check

leofang commented:
Note from meeting: @MrRobot2211 would be great to post your benchmark script.

MrRobot2211 commented:
> @hodegestar I have been doing some benchmarking and completing the rest of the functions in the other PR. The aggressive calls to `cp.array`, through the calls to `__init__` in the methods, produce a time penalty of around 1.7X when doing some large matrix multiplications. I have two options to circumvent this or I can define methods like this:
> 
> ```
> def adjoint(self):
>         
>         cp_out = self._cp.transpose().conj()
>        
>         return build_cpd_from_cp(cp_out)
> 
> def  build_cpd_from_cp( data):
>       out = CuPyDense.__new__(CuPyDense)
>       super(CuPyDense , out).__init__(data.shape)
>       out._cp = data
>       return out
> ```
> 
> ( which avoids the shape checks, but this new `build_cpd_from_cp` constructor seems awkward)
> or by adding a type check in the `__init__`
> 
> ```
> def __init__(self, data, shape=None, copy=True):
>         if isinstance(data, cp.ndarray):
>             shape = data.shape
>             self._cp = data
> ```
> 
> What are your thoughts here?

While looking at this I realized that @jakelishman  might have been pointing outt the fact that `copy = True` s in the default initialization. Which is in fact where the highest penalization occurs, besides the call to cupy.array.

As I understand it the way of solving only that is to keep the defaults as they are for compatibility, and just change the argument when calling the class methods. But this still does incur a minor penalization of calling `cupy.array`.

MrRobot2211 commented:
> Note from meeting: @MrRobot2211 would be great to post your benchmark script.

Thank you @leofang, I am uploading them later on today.

jakelishman commented:
> While looking at this I realized that @jakelishman might have been pointing outt the fact that `copy = True` s in the default initialization. Which is in fact where the highest penalization occurs, besides the call to cupy.arr

Good job, yeah, that's the most major thing I'd seen.  In general, remember that at every point in internal operations, you should know who the owner of a particular array is meant to be.  If it's new, and you know you're just constructing it "for" something else and you'll not keep any references to it (like here), then you shouldn't copy the data unnecessarily.  The default to `__init__` must be `copy=True`, because that's the safe option; you can't know if a user will have multiple references to a single `cupy.ndarray`, so it's not safe for you to just wrap it, in case they change one of their views.  They have to positively _tell_ you otherwise for it to be safe.  If you look at various functions which create `Qobj` in `dev.major`, you'll see the same pattern - the data object is created on its own, and then it's passed to `Qobj` with a `copy=False` argument.

You may also want to consider if your calls to `cupy.array` might create overhead.  In particular, consider the effects of the major ordering, and if unnecessary conversions could ever occur.  If you find that even in the fastest cases (and calls to `cupy.array` might not be the fastest way to achieve what you need here) there's still significant overhead, then consider if you can implement a "fast path" through your initialiser for internal use (one way could be a separate, private class-method constructor which does less error checking).

hodgestar commented on src/qutip_cupy/__init__.py:
```diff
@@ -1 +1,25 @@
 from .version import version as __version__
+from .cupy_dense import *
```
I would not use `import *` here because it hinders static analysis tools like `pyflakes`. I would do something like:
```python
import .cupy_dense as cd

CuPyDense = cd.CuPyDense

...

del cd
del data
```

leofang replied:
How about just `from .cupy_dense import CuPyDense`?

hodgestar commented on src/qutip_cupy/__init__.py:
```diff
@@ -1 +1,25 @@
+from .cupy_dense import *
+
+
+from qutip.core import data
```
Put all imports together at the top of the file (unless they're inside functions or methods, e.g. as an optional dependency).

For extra credit, divide the imports into three paragraphs -- one for modules in the standard library, one for third party modules, and one for modules from this package.

hodgestar commented on src/qutip_cupy/__init__.py:
```diff
@@ -1 +1,25 @@
 from .version import version as __version__
```
I would add a module docstring right at the top of the file, e.g.

```
""" The qutip-cupy package provides a CuPy-based data layer for QuTiP. """
```

hodgestar commented on src/qutip_cupy/cupy_dense.py:
```diff
@@ -0,0 +1,88 @@
+import cupy as cp
```
Could we also add a short module docstring here?

hodgestar replied:
I would also call this file simply `dense.py` since the package itself already mentions `cupy`.

hodgestar commented on src/qutip_cupy/cupy_dense.py:
```diff
@@ -0,0 +1,88 @@
+from qutip.core import data
+
+
+class CuPyDense(data.Data):
```
The `CuPyDense` class docstring should detail the arguments to `__init__` and how `CuPyDense` should be used and have a very short explanation to how the data class works under the hood (e.g.  it stores data in arrays on the GPU using CuPy, it copies data in these circumstances, etc).

hodgestar commented on src/qutip_cupy/cupy_dense.py:
```diff
@@ -0,0 +1,88 @@
+        return CuPyDense(self._cp.copy())
+
+    def to_array(self):
+        """ 
```
This docstring can be removed, I think. As discussed in the call today `help(CuPyDense.to_array)` will read the docstring from `Data.to_array`.

MrRobot2211 replied:
In this case `Data.to_array` does not have any doctring,and for that matter most of the methods of `data.Data` since they are not "implemented" .  Besides the behaviour is different here, we are not only copying to a numpy array  we are migrationg  the data from the GPU to the CPU which is a bigger penalty and the user should be made aware.

hodgestar commented on src/qutip_cupy/cupy_dense.py:
```diff
@@ -0,0 +1,88 @@
+        return self._cp.trace()
+
+
+def dense_from_cupydense(cupydense):
```
These methods should have docstrings describing the types of their inputs and what they do (they'll likely be quite short). They'll be especially important if the specializations start supporting multiple types (e.g. single and double precision complex CuPy arrays).

hodgestar commented on tests/test_base_dense.py:
```diff
@@ -0,0 +1,49 @@
+from qutip_cupy import CuPyDense
```
I would name this file `test_dense.py` to match `dense.py`. In general, most test files should have the name `test_<module-being-tested>.py` although there will be exceptions at times (e.g. maybe we will want to test what is exposed by a package).

hodgestar replied:
I would perhaps also organize the tests in the file into classes since there might be a lot of them, e.g.
```python

class TestCuPyDense:
    ... tests of the methods and constructor of CuPyDense go here ...

class TestAdjointSpecialisation:
    ... adjoint tests go here ...

class TestConjSpecialisation:
    ...
```
How exactly to organize things is up for discussion, but it would be good to have some ordering.

hodgestar commented on tests/test_linalg.py:
```diff
@@ -0,0 +1,15 @@
+import time
```
I think we can delete this file for now?

hodgestar reviewed (changes requested):
@MrRobot2211 This is looking really nice so far! I left quite a few comments nitpicking the code, but don't let it distract from the fact that you're making really nice progress.

hodgestar commented:
@MrRobot2211 Woot on finding the `copy=True` issue! Note that `cp.array` *also* has a `copy=True` default that you could override by passing `copy=copy` in the `CuPyDense` constructor, which might save you another copy of the data on the GPU.

MrRobot2211 commented:
> @MrRobot2211 Woot on finding the `copy=True` issue! Note that `cp.array` _also_ has a `copy=True` default that you could override by passing `copy=copy` in the `CuPyDense` constructor, which might save you another copy of the data on the GPU.

@hodgestar If I am not mistaken it is already being overriden that way so `copy=False` in `CuPyDense` means `copy=False` in `cp.array` at i`__init__`. But  there is  still a small penalty on just calling `cp.array`( with `copy=False` set ) which I'll probably just skip by calling an specialized constructor.

hodgestar commented:
> @hodgestar If I am not mistaken it is already being overriden that way so `copy=False` in `CuPyDense` means `copy=False` in `cp.array` at i`__init__`. But there is still a small penalty on just calling `cp.array`( with `copy=False` set ) which I'll probably just skip by calling an specialized constructor.

A specialized constructor might also be good (and could allow skipping some shape tests too, I guess).

A common pattern for these is to use a classmethod and perhaps some special flag to the constructor (e.g. `_i_know_what_i_am_doing_and_have_a_cupy_array_and_the_shape_is_right=True`).

hodgestar commented:
> A common pattern for these is to use a classmethod and perhaps some special flag to the constructor (e.g. `_i_know_what_i_am_doing_and_have_a_cupy_array_and_the_shape_is_right=True`).

Or you can go down this rabbithole -- https://stackoverflow.com/questions/2168964/how-to-create-a-class-instance-without-calling-initializer/. I don't really recommend most of those for actual production code, but they are fun!

MrRobot2211 committed:
adds specialized constructor

MrRobot2211 committed:
renames cupy_dense

MrRobot2211 committed:
Refactors __init__
adds docstring, dense deletion after use and sorts imports properly

MrRobot2211 committed:
deletes test_linalg

MrRobot2211 committed:
adds docstrings to  cupydense class

jakelishman commented:
> Or you can go down this rabbithole -- https://stackoverflow.com/questions/2168964/how-to-create-a-class-instance-without-calling-initializer/. I don't really recommend most of those for actual production code, but they are fun!

The Cython implementations of `Dense` and `CSR` actually do this type of thing _a lot_, because their `__init__` methods are intended for Python-space, whereas most of their constructions are done at the C level.  I think overriding `__new__` is a massive can of worms (and manually setting `x.__class__` is horrendous...), but a specialised constructor not calling `__init__`, only `__new__`, is a pretty safe pattern (assuming you're not dealing with crazy inheritance).

MrRobot2211 commented:
@leofang  I left some benchmarks [here](https://github.com/MrRobot2211/qutip-cupy/blob/benchmarks/benchmarks/mul_and_cparray_calls_benchmark.ipynb) , these are on a 1080ti with cudatoolkit 10.1 . I am doing the same ones on a 2080ti that I have on a faster lane. 

I will be adding some `qutip_cupy` in-library benchmarks later on as well.

MrRobot2211 committed:
Merge branch 'main' of https://github.com/qutip/qutip-cupy into dense_no_cython

Here goes nothing !

aktech commented:
@hodgestar @MrRobot2211

Error from AWS, when creating a new runner.

```
An error occurred (VcpuLimitExceeded) when calling the RunInstances operation: You have requested more vCPU capacity than your current vCPU limit of 8 allows for the instance bucket that the specified instance type belongs to. Please visit http://aws.amazon.com/contact-us/ec2-request to request an adjustment to this limit.
```
(Two G instances is the max you can run with your current AWS limit.)

The reason for the latest build being stuck is potentially the two instances that were created for two jobs earlier, were being deleted and during that time your vCPU limit is not reset and you requested another instance and that's why the error above.

TL;DR for these kind of issues: Just trigger another build, via a new commit in a bit.

hodgestar commented:
@aktech That's a bit odd. I thought AWS approved the limit increase to 64 vCPUs, but maybe that hasn't come through yet.

aktech commented:
Yeah, It takes a while sometimes to get limits increased in practice.

MrRobot2211 committed:
removes unneecessary cpp exclusion

MrRobot2211 committed:
silences qutip matplotlib warning

MrRobot2211 committed:
silences np.bool warning from cupy

MrRobot2211 committed:
changes cupy conda channel
this makes the installation be the latest released
and avoids silencing the np.bool
deprecation warning in cupy

MrRobot2211 committed:
fixes typo in doc  cupy channel

MrRobot2211 committed:
adds device selection for CuPyDense
to the docstring

MrRobot2211 committed:
adds dtype attribute

MrRobot2211 committed:
added doctring for conversion functions

MrRobot2211 committed:
renames test_dese

MrRobot2211 committed:
fixes typo in docstring

MrRobot2211 committed:
refactors test _dense into classes

MrRobot2211 committed:
Atest: adds transpose test and changes assert
to np.testing._assert where necessary

MrRobot2211 committed:
changes calls to cls inclassmethod
instead of  CuPyDense

MrRobot2211 committed:
tests inheritance of CuPyDense methods

AGaliciaMartinez commented on tests/test_dense.py:
```diff
@@ -55,3 +55,15 @@ def test_trace(self, matrix, trace):
+
+
+def test_no_checks_constr():
+    """Checks if CuPyDense canbe properly inherited"""
```
```suggestion
    """Checks if CuPyDense can be properly inherited"""
```

hodgestar replied:
I'm not really sure what this is testing? I.e. in what circumstances do we expect to wrap ``CuPyDense`` and how might it fail?

E.g. I would never expect the code below to fail because it's only using basic Python semantics -- this is sub-class of a pure Python class testing that the inherited methods exist.

There are things that I would expect to fail -- e.g. ``wrapcpd`` hasn't been registered with the data layer, but I think we should just delete this test for now and revisit it when we have something more concrete to test.

MrRobot2211 replied:
This test was purely to debug the custom constructor in case we needed some kind of inheritance, I am erasing it.

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,130 @@
+
+class CuPyDense(data.Data):
+    def __init__(self, data, shape=None, copy=True, dtype=cp.complex128):
+        """ 
```
This docstring for ``__init__`` should be on the class itself (this is convention, but also makes a lot of sense since it usually also contains documentation on how to use the class). E.g.
```python
class CuPyDense(data.Data):
    """
    This class provides a dense matrix ...
    """

    def __init__(self, data, ...):
        ...
```

MrRobot2211 replied:
423ac6a

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,130 @@
+            else:
+                self._cp = base.reshape(shape)
+        else:
+
```
```suggestion
```

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,130 @@
+        return self._no_checks_constructor(self._cp.copy())
+
+    def to_array(self):
+        """ 
+        Get a copy as a `numpy.ndarray`.
+        This incurs memory-transfer from `host` to `device`
+        and should be avoided when possible    
+        """
```
```suggestion
```
I know this isn't documented in qutip yet, but I still think we should remove the docstring here and rather document it in qutip itself.

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,130 @@
+        super().__init__((shape[0], shape[1]))
+
+    @classmethod
+    def _no_checks_constructor(cls, data):
+        # This constructor should only be called when we are sure the argument
+        # is a CuPy 2-dimensional array.
```
```suggestion
    def _raw_cupy_constructor(cls, data):
        """
        A fast low-level constructor for wrapping an existing CuPy array in a CuPyDense object 
        without copying it.   
        
        The ``data`` argument must be a CuPy array with the correct shape.
        The CuPy array will not be copied and will be used as is. 
        """
```
I have a personal preference for putting important information into docstrings rather than comments when it's about a function as a whole.

I have a hunch that we'll want to expose this alternative constructor publicly at some point, but we'll need a better name for it and let's not do it now (we can do it later if it still seems like a good idea). The hunch is that it's very useful for us, so it will likely be very useful to someone else who has a CuPy array and knows what they're doing.

MrRobot2211 replied:
b23aec8

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,130 @@
+# I am not really sure how the dispatcher would handle it.
+# It looks like we may be needing 2 classes. 
+def dense_from_cupydense(cupydense):
+    """"Creates a QuTiP ``data.Dense`` array from the values in a CuPyDense array.
+        The resulting array has complex128 precision. """
```
```suggestion
    """
    Creates a QuTiP ``data.Dense`` array from the values in a CuPyDense array.
    The resulting array has complex128 precision.
    """
```

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,130 @@
+
+
+def cupydense_from_dense(dense):
+    """ Creates a CuPyDense array from the values in a QuTiP ``data.Dense`` array 
+        with ``cp.complex128`` precision. """
```
```suggestion
    """
    Creates a CuPyDense array from the values in a QuTiP ``data.Dense`` array 
    with ``cp.complex128`` precision.
    """
```

hodgestar commented on src/qutip_cupy/__init__.py:
```diff
@@ -1 +1,43 @@
+from . import dense as cd
+CuPyDense = cd.CuPyDense
+
+data.to.add_conversions([
+     (CuPyDense, data.Dense, cd.cupydense_from_dense),
+     (data.Dense, CuPyDense, cd.dense_from_cupydense),
+ ])
+data.to.register_aliases(['cupyd'], CuPyDense)
+
+data.adjoint.add_specialisations([
+      (CuPyDense, CuPyDense, cd.cpd_adjoint),
+     ])
+data.transpose.add_specialisations([
+      (CuPyDense, CuPyDense, cd.cpd_transpose),
+     ])
+data.conj.add_specialisations([
+      (CuPyDense, CuPyDense, cd.cpd_conj),
+     ])
```
```suggestion
data.to.add_conversions([
    (CuPyDense, data.Dense, cd.cupydense_from_dense),
    (data.Dense, CuPyDense, cd.dense_from_cupydense),
])
data.to.register_aliases(['cupyd'], CuPyDense)

data.adjoint.add_specialisations([
    (CuPyDense, CuPyDense, cd.cpd_adjoint),
])
data.transpose.add_specialisations([
    (CuPyDense, CuPyDense, cd.cpd_transpose),
])
data.conj.add_specialisations([
    (CuPyDense, CuPyDense, cd.cpd_conj),
])
```
I'm trying to tidy up the indentation a bit -- perhaps we can introduce one of the opinionated formatters like ``black`` or ``blue`` to avoid these sorts of review comments. :)

MrRobot2211 replied:
I have pcodestyle, but till now I was just ignoring it, should have started paying attentionbefore. What do you think of adding a style check to the CI?

hodgestar replied:
I would be very much in favour of adding a style check to the CI. My vote is for pyflakes and black.

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,130 @@
+    return cpd_array.transpose()
+
+
+def cpd_trace(cpd_array):
```
We define this function but we don't register it with the specialization yet.

MrRobot2211 replied:
53c5323

hodgestar commented on tests/test_dense.py:
```diff
@@ -0,0 +1,69 @@
+    """ Tests of the methods and constructors of the CuPyDense class. """
+
+    def test_shape(self, shape):
+
+        cupy_dense = CuPyDense(np.random.uniform(size=shape))
+
+        assert (cupy_dense.shape == shape)
```
```suggestion
        cupy_dense = CuPyDense(np.random.uniform(size=shape))
        assert (cupy_dense.shape == shape)
```
No need to have blank lines in short functions -- they just make the file longer and harder to read, I think.

hodgestar commented on tests/test_dense.py:
```diff
@@ -0,0 +1,69 @@
+        assert (cupy_dense.shape == shape)
+
+    def test_transpose(self, shape):
+
+        cupy_dense = CuPyDense(np.random.uniform(size=shape)).transpose()
+        
+        np.testing.assert_array_equal(cupy_dense.shape, (shape[1],shape[0]))
```
```suggestion
        data = np.random.uniform(size=shape)
        cd = CuPyDense(data).transpose()
        assert cd.shape == (shape[1], shape[0])
        np.testing.assert_array_equal(cd.to_array(), data.transpose())
```
We also need to assert that the transposed dense array has the correct numbers in it!

MrRobot2211 replied:
np.testing.assert_array_equal  as per description checks the equality of the shapes already.

MrRobot2211 replied:
unless we want a separate check on the shape to be sure it it not a failure in copying the values.

hodgestar replied:
I'm happy for us to delete the assert on the shape.

hodgestar commented on tests/test_dense.py:
```diff
@@ -0,0 +1,69 @@
+        np.testing.assert_array_equal(cupy_dense.shape, (shape[1],shape[0]))
+
+    def test_adjoint(self, shape):
+
+        from qutip.core import data
+
+        array = np.random.uniform(size=shape) + 1.j*np.random.uniform(size=shape)
+
+        cpdense_adj = CuPyDense(array).adjoint()
+        qtpdense_adj = data.Dense(array).adjoint()
+
+        np.testing.assert_array_equal(cpdense_adj.to_array(), qtpdense_adj.to_array())
```
```suggestion
        data = np.random.uniform(size=shape) + 1.j*np.random.uniform(size=shape)
        cp_adj = CuPyDense(data).adjoint()
        np.testing.assert_array_equal(cpdense_adj.to_array(), data.transpose().conj())
```
No need to involve `data.Dense` here -- keep the tests simple.

hodgestar commented on tests/test_dense.py:
```diff
@@ -0,0 +1,69 @@
+    @pytest.mark.parametrize(["matrix", "trace"], [pytest.param([[0, 1],[1, 0]], 0),
+                                                pytest.param([[2.j, 1],[1, 1]], 1+2.j)])
+    def test_trace(self, matrix, trace):
+
+        cupy_array = CuPyDense(matrix)
+
```
```suggestion
        cupy_array = CuPyDense(matrix)
```
More blank line removal.

hodgestar reviewed (changes requested):
@MrRobot2211 I think after this round of comments we should be ready to land this branch. In the interests of speeding things up I made most of my comments as suggestions. Please commit the ones you like and write a response to the ones you don't, and then I'll do a quick last check and hopefully we can merge after that.

Looking good!

MrRobot2211 committed:
Removes extra line from .gitignore

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
Adds spacing after docstring in __init__

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
removes blank_line in dense

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
removes extra blank lines from __init__

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
refactor adjoint test

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
- blank line in test dense

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
- blank line in dense

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
= blank line  in gitignore

MrRobot2211 commented on setup.cfg:
```diff
@@ -29,6 +29,7 @@ packages = find:
 include_package_data = True
 install_requires =
     qutip>=5.0.0.dev0
+    cupy
```
@hodgestar related but not necesarily linked to this. Should we remove test_base that tests if cupy has been installe or are we keeping it to check there are no failures of the environment in the CI? If so I should erase the header line in test_base.

leofang replied:
Yeah wrapping `import CuPy` in a try-except block makes sense.

hodgestar replied:
> @hodgestar related but not necessarily linked to this. Should we remove test_base? it tests if cupy has been installe or are we keeping it to check there are no failures of the environment in the CI? If so I should erase the header line in test_base.

Good question. Removing it is one option. Renaming it to `test_00_sanity_check_environment.py` or similar is not a bad idea either. That way one can immediately see if GPUs are missing or something else vital.

Up to you.

MrRobot2211 commented on src/qutip_cupy/.gitignore:
```diff
@@ -1 +1 @@
-version.py
+version.py
```
fixed in fa3b90f

MrRobot2211 committed:
adds trace to the dispatcher

MrRobot2211 committed:
lints tests

MrRobot2211 committed:
moves CuPyDense docstring

MrRobot2211 committed:
corrects dense docstring

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
renamed no_checks_constructor to raw
 and adds docstring

MrRobot2211 committed:
fixes docstring lines

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
fixes docstring lines

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
removes  docstring for to_array

this function should be documented in qutip

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
fixes extra space in init

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
fixes white spaces in init

MrRobot2211 committed:
lints

MrRobot2211 committed:
test: rmoves inheritance on dense check

MrRobot2211 committed:
renames test_base to environment sanity check

MrRobot2211 commented:
Well I think that's the last one @hodgestar :)

MrRobot2211 committed:
fixes test name typo

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,135 @@
+
+class CuPyDense(data.Data):
+    """
+        This class provides a dense matrix backend for QuTiP.
+        Matrices are stored internally in a CuPy array on a GPU.
+        If you have many GPUs you can set GPU ``i``
+        by calling ``cp.cuda.Device(i).use()`` before construction.
+
+        Parameters
+        ----------
+        data: array-like
+            Data to be stored.
+        shape: (int, int)
+            Defaults to ``None``. If ``None`` will infer the shape from ``data``,
+            else it will set the shape for the internal CuPy array.
+        copy: bool
+            Defaults to ``True``. Whether to make a copy of
+            the elements in ``data`` or not.
+        dtype:
+            Data type specifier. Either ``cp.complex128`` or ``cp.complex64``
+
```
```suggestion
    This class provides a dense matrix backend for QuTiP.
    Matrices are stored internally in a CuPy array on a GPU.
    If you have many GPUs you can set GPU ``i``
    by calling ``cp.cuda.Device(i).use()`` before construction.

    Parameters
    ----------
    data: array-like
        Data to be stored.
    shape: (int, int)
        Defaults to ``None``. If ``None`` will infer the shape from ``data``,
        else it will set the shape for the internal CuPy array.
    copy: bool
        Defaults to ``True``. Whether to make a copy of
        the elements in ``data`` or not.
    dtype:
        Data type specifier. Either ``cp.complex128`` or ``cp.complex64``
```
Properly indent docstring.

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,135 @@
+    """
+
+    def __init__(self, data, shape=None, copy=True, dtype=cp.complex128):
+
```
```suggestion
```

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -0,0 +1,135 @@
+    Creates a QuTiP ``data.Dense`` array from the values in a CuPyDense array.
+    The resulting array has complex128 precision.
+    """
+
```
```suggestion
```

hodgestar commented on tests/test_dense.py:
```diff
@@ -0,0 +1,47 @@
+        assert (cupy_dense.shape == shape)
+
+    def test_transpose(self, shape):
+
+        cupy_dense = CuPyDense(np.random.uniform(size=shape)).transpose()
+
+        np.testing.assert_array_equal(cupy_dense.shape, (shape[1], shape[0]))
```
```suggestion
        cupy_dense = CuPyDense(np.random.uniform(size=shape)).transpose()
        np.testing.assert_array_equal(cupy_dense.shape, (shape[1], shape[0]))
```

hodgestar committed:
Formatting fixes.

hodgestar commented:
@MrRobot2211 Nice work! I cleaned up a couple more formatting things and merged.

I opened #12 to remind us to address the `cupy` dependency in `setup.cfg` later.
